src/screencast/screencast.ts

```diff
@@ -0,0 +1,193 @@
+        this.urlInput = document.getElementById('url') as HTMLInputElement;
+        this.screencastImage = document.getElementById('canvas') as HTMLImageElement;
+        this.screencastWrapper = document.getElementById('canvas-wrapper') as HTMLElement;
+        this.deviceSelect = document.getElementById('device') as HTMLSelectElement;
+
+        this.backButton.addEventListener('click', () => this.onBackClick());
+        this.forwardButton.addEventListener('click', () => this.onForwardClick());
+        this.reloadButton.addEventListener('click', () => this.onReloadClick());
+        this.rotateButton.addEventListener('click', () => this.onRotateClick());
+        this.urlInput.addEventListener('keydown', event => this.onUrlKeyDown(event));
+
+        this.deviceSelect.addEventListener('change', () => {
+            if (this.deviceSelect.value.toLowerCase() === 'desktop') {
+                this.width = 0;
+                this.height = 0;
+                this.screencastWrapper.classList.add('desktop');
+            } else {
+                const selectedOption = this.deviceSelect[this.deviceSelect.selectedIndex];
+                const deviceWidth = selectedOption.getAttribute('devicewidth');
+                const deviceHeight = selectedOption.getAttribute('deviceheight');
+                if (deviceWidth && deviceHeight) {
+                    this.width = parseInt(deviceWidth);
+                    this.height = parseInt(deviceHeight);
+                }
+
+                this.screencastWrapper.classList.remove('desktop');
+            }
+            this.updateEmulation();
+        });
+
+        this.cdpConnection.registerForEvent('Page.domContentEventFired', () => this.onDomContentEventFired());
+        this.cdpConnection.registerForEvent('Page.frameNavigated', result => this.onFrameNavigated(result));
+        this.cdpConnection.registerForEvent('Page.screencastFrame', result => this.onScreencastFrame(result));
+
+        this.inputHandler = new ScreencastInputHandler(this.cdpConnection);
+
+        this.cdpConnection.sendMessageToBackend('Page.enable', {});
+        let resizeTimeout = 0 as unknown as NodeJS.Timeout;
+        window.addEventListener('resize', () => {
+            clearTimeout(resizeTimeout);
+            resizeTimeout = setTimeout(() => this.updateEmulation(), 100);
+        });
+
+        this.registerInputListeners();
+        
+        // Start screencast
+        this.updateEmulation();
+        this.updateHistory();
+    }
+
+    private registerInputListeners(): void {
+        for (const eventName of Object.keys(MouseEventMap)) {
+            this.screencastImage.addEventListener(eventName, (event) => {
+                const scale = this.screencastImage.offsetWidth / this.screencastImage.naturalWidth;
+                if (this.isDeviceTouch()) {
+                    this.inputHandler.emitTouchFromMouseEvent(event as MouseEvent, scale);
```

Comment:
Currently this keeps "Inspect Element" from the DevTools from highlighting nodes as we hover over them in the preview when emulating a touch device. We should try to send mouse events while the inspect mode is active to allow the overlay to work.

CDP at least tells us when the mode has been cancelled (https://chromedevtools.github.io/devtools-protocol/tot/Overlay/#event-inspectModeCanceled) but I'm not as sure about a CDP event for it being activated (we may have to forward something from the tools).

Reply:
Discussed this offline with @antross, as the new behavior will be off by default (while we stabilize it) we will follow up with this one on another PR to make it easier to review.